payment-guarantee/src/msg.rs

```diff
@@ -0,0 +1,27 @@
+use cosmwasm_std::Coin;
+use cw721::Cw721ReceiveMsg;
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+
+#[derive(Serialize, Deserialize, Clone, PartialEq, JsonSchema, Debug)]
+pub struct InstantiateMsg {
+    pub source_contracts: Vec<String>,
+}
+
+
+#[derive(Serialize, Deserialize, Clone, PartialEq, JsonSchema, Debug)]
+#[serde(rename_all = "snake_case")]
+pub enum ExecuteMsg {
+    Deposit {},
+    ReceiveNft(Cw721ReceiveMsg),
+}
+
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+#[serde(rename_all = "snake_case")]
+pub struct TokenInfoMsg {
+    pub contract: String,
+    pub description: Option<String>,
+    pub price: Coin,
+    pub sender: String,
+}
```

Comment:
It seems that the `contract` and the `description` aren't used anywhere.

Reply:
That's right. I'm not using it right now.
The current development is minimal verification and exchange.
`contract` will be used for verification.
Description will be shown as described in NFT when showing contract status.(with token_id, name, etc...)

Reply:
In addition, Successful NFT transactions are achieved on NFT transmission (validated and exchanged immediately).

I'm thinking about what action to take to exchange the NFT.
* When NFT was sent (currently implemented)
* Alternatively, perform verification and exchange operations in a separate action.

Reply:
Is the `TokenInfoMsg` only for the `Cw721ReceiveMsg.msg`? Then, only the `sender` would be enough.

I'm confused what's the difference with `TokenInfoMsg` and `TokenInfo` (in the PR #3). I saw that the `TokenInfo` is put into the `cw721-base.TokenInfo.description` as JSON. That's good.
Then, I guess we can extract the `cw721-base.TokenInfo.description` when we handle the `Cw721ReceiveMsg`.
Then, we can get `contract`, `description` and `price` from it.

I'm not sure if it's possible, but it looks much simpler because we don't need any duplicated declarations.

Reply:
TokenInfo, located in panacea-nft, is used to store additional information when MINT the NFT.
As a result, if only a `sender` is added, it is the same structure as TokenInfoMsg.
The `sender` is only required to perform SEND_NFT, so it is not included in the Structures field used to save.

Since panacea-nft and payment-guarantee are different projects, declarations on structures have been made respectively.
- In `panacea-nft`, declare to `TokenInfo` and `TokenInfoWithSender`
- In `payment-guarantee`, declar to `TokenInfoMsg`.

I think 'TokenInfoMsg' should be used to declare a common structure. 
However, we can distribute it to crates.io and manage it or declare it as below.
`payment-guarantee = { path = "../payment-guarantee", version = "0.0.1" }`

Reply:
Let me rephrase the topic.

In my guess,
In `payment-guarantee`, the `TokenInfoMsg` can be defined as below:
```
pub struct TokenInfoMsg {
    pub sender: String,
}
```
Also, we can rename `TokenInfoMsg` to `ReceiveNftMsgBody` or something else, because that isn't the token info. That's the additional info for handling the `ReceiveNft` operation.

If you need the token info when handling the `ReceiveNft` operation, I think we can get the `TokenInfo` (defined in the `panacea-nft`) using `deps.querier.query_wasm_smart()`.
I didn't try it yet, but it seems it's feasible.

Then, as you said, we should do:
```
panacea-nft = { path = "../panacea-nft", version = "0.0.1" }
```
(IMHO, the `panacea-nft` is the mother project of the `payment-guarantee`, not the other way around.)

Reply:
My main concern is that it looks inflexible to declare duplicated fields (`contract`, `description` and `price`) in both of `TokenInfo` and `TokenInfoMsg`.

Reply:
And, please put `TODO` comments about what further validations should be added in the future.

Reply:
Aha, I understand. 
Likewise, I also thought about duplicate fields.
The method you suggested is an implementable method.

When inquiring in `deps.querier` to verify that the NFT is contract-owned, I think it would be better to other information be imported and used together.(set price)

Reply:
> And, please put `TODO` comments about what further validations should be added in the future.
I'm writing what needs to be developed and validation here.

https://github.com/medibloc/panacea-smart-contract/issues/5